### PR TITLE
Bug 1878735 - Ingest Yelp icon from Remote Settings

### PR DESCRIPTION
**Pull request by ncloudioj**
### Pull Request checklist ###

- **Breaking changes**:  This PR follows our [breaking change policy](https://github.com/mozilla/application-services/blob/main/docs/howtos/breaking-changes.md)
  - [ ] This PR follows the breaking change policy:
     - This PR has no breaking API changes, or
     - There are corresponding PRs for our consumer applications that resolve the breaking changes and have been approved
- [ ] **Quality**: This PR builds and tests run cleanly
  - Note:
    - For changes that need extra cross-platform testing, consider adding `[ci full]` to the PR title.
    - If this pull request includes a breaking change, consider [cutting a new release](https://github.com/mozilla/application-services/blob/main/docs/howtos/cut-a-new-release.md) after merging.
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes a changelog entry in [CHANGELOG.md](../CHANGELOG.md) or an explanation of why it does not need one
  - Any breaking changes to Swift or Kotlin binding APIs are noted explicitly
- [ ] **Dependencies**: This PR follows our [dependency management guidelines](https://github.com/mozilla/application-services/blob/main/docs/dependency-management.md)
  - Any new dependencies are accompanied by a summary of the due dilligence applied in selecting them.

[Branch builds](https://github.com/mozilla/application-services/blob/main/docs/howtos/branch-builds.md): add `[firefox-android: branch-name]` to the PR title.

### CODE REVIEW

File: components/suggest/src/schema.rs [hunk 1]
```diff
@@ -92,6 +92,11 @@ pub const SQL: &str = "
         record_id TEXT NOT NULL
     ) WITHOUT ROWID;
 
+    CREATE TABLE yelp_custom_details(
```
Comment (ncloudioj) [inline components/suggest/src/schema.rs:95]: Unlike other suggestions, we build Yelp suggestions on the fly, so there is no records in the `suggestions` table for Yelp. Thus, we don't have the `suggestion_id` in this table. Also, there should be only one record in this table, which feels odd. As discussed in another thread, maybe we can move the `icon_id` to the `suggestion_providers` table in the future.

File: components/suggest/src/db.rs [hunk 2]
```diff
@@ -1071,6 +1071,10 @@ impl<'a> SuggestDao<'a> {
             "DELETE FROM yelp_location_signs WHERE record_id = :record_id",
             named_params! { ":record_id": record_id.as_str() },
         )?;
+        self.conn.execute_cached(
+            "DELETE FROM yelp_custom_details WHERE record_id = :record_id",
+            named_params! { ":record_id": record_id.as_str() },
+        )?;
```
Comment (0c0w3) [inline components/suggest/src/db.rs:1077]: Does the icon data in `icons` need to be deleted too? Does that happen somewhere else?

**Review by 0c0w3 — APPROVED** [review]
Looks good, thanks again for doing this!

Repeating the icon bytes in the test is kind of too bad -- would it be reasonable to factor it out into a const or something?

File: components/suggest/src/db.rs (hunk 2 above)
Comment (ncloudioj) [reply components/suggest/src/db.rs:1077]: Yup, icon deletion happens [here](https://github.com/mozilla/application-services/blob/a757c1ec4083259622e366ea4ce5a2c2189b2088/components/suggest/src/db.rs#L167).

**Comment by ncloudioj** [comment]
> Repeating the icon bytes in the test is kind of too bad -- would it be reasonable to factor it out into a const or something?

I know, it's kinda annoying. But the entire component is stateless, so I couldn't find a better way to cache that to avoid repeated loads.

@linabutler @bendk Wonder if you have better recommendations?

**Comment by bendk** [comment]
> > Repeating the icon bytes in the test is kind of too bad -- would it be reasonable to factor it out into a const or something?
> 
> I know, it's kinda annoying. But the entire component is stateless, so I couldn't find a better way to cache that to avoid repeated loads. This also happens to Wikipedia, we store 7,000 icon ids (they are identical) in `wikipedia_custom_details`! And we load the icon from db for every single hit of Wikipedia.
> 
> @linabutler @bendk Wonder if you have better recommendations?

I think you're talking about two different things.

Drew's comment seems to be about these [lines in the tests](https://github.com/mozilla/application-services/blob/0b51ee950b799f1a4c193aa0ad7d4d1eee855a8f/components/suggest/src/store.rs#L816-L830).  Could that `icon` value be factored out into a const at the top to avoid the repetition?

Nan's comment seems to be about the fact that icon data might be loaded multiple times, even though it's easily cachable.  I guess we could start storing some cache in the component somewhere, but it doesn't seem that high-priority to me.  My guess is that SQLite + the disk cache will be fast enough.

**Comment by ncloudioj** [comment]
> I think you're talking about two different things.

Oops, sorry, I misread Drew's comment. Those test fixtures are automatically populated and managed by [expect_test](https://docs.rs/expect-test/latest/expect_test/). So I don't think we can swap in some constants to minimize those unfortunately.

> I guess we could start storing some cache in the component somewhere, but it doesn't seem that high-priority to me. My guess is that SQLite + the disk cache will be fast enough.

Agreed, no need to worry too much right now. Just be mindful about the icon size (~10KB across all the providers), repeatedly loading those from the db might be sub-optimal.

**Review by dadaa — APPROVED** [review]
Thank you very much, @ncloudioj! LGTM!

File: components/suggest/src/suggest.udl [hunk 3]
```diff
@@ -73,7 +73,8 @@ interface Suggestion {
     Yelp(
         string url,
         string title,
-        boolean is_top_pick
+        boolean is_top_pick,
+        sequence<u8>? icon
```
[the remark this answers is not in the document]
Comment (ncloudioj) [reply components/suggest/src/suggest.udl:77]: Yes, I think so. While the `icon` is a required field for `DownloadedYelpAttachment`, it's just an ID pointing to the icon record and we can't guarantee that icon's present on RS (e.g. someone could delete it by accident).

**Pull request merged**